### PR TITLE
Add user service

### PR DESCRIPTION
**Pull request by kvtim**
Create user service (Authentication, authorization, CRUD, change password)

### CODE REVIEW

File: CryptoExchange/Microservices/CurrencyManagement/CurrencyManagement.Api/Controllers/CurrenciesController.cs [hunk 1]
```diff
@@ -28,12 +27,33 @@ public async Task<IActionResult> GetAll()
             return Ok(_mapper.Map<IEnumerable<CurrencyDto>>(currencies));
         }
 
+        [HttpGet("AllWithDimension")]
+        public async Task<IActionResult> GetAllWithDimension()
+        {
+            var currencies = await _currencyService.GetAllWithDimensionAsync();
+            return Ok(_mapper.Map<IEnumerable<CurrencyWithDimensionDto>>(currencies));
```
Comment (Vlad1slav-Hmeln1tsk1j) [inline CryptoExchange/Microservices/CurrencyManagement/CurrencyManagement.Api/Controllers/CurrenciesController.cs:34]: If currencies will be null that's anyway will be ok?

File: CryptoExchange/Microservices/CurrencyManagement/CurrencyManagement.Api/Controllers/CurrenciesController.cs [hunk 2]
```diff
@@ -43,8 +63,7 @@ public async Task<IActionResult> Add([FromBody] CreateCurrencyDto createCurrency
 
             var currencyResult = await _currencyService.AddAsync(currency);
 
-            return currency == null ? BadRequest() : Ok(_mapper.Map<CurrencyDto>(currencyResult));
-
+            return Ok(_mapper.Map<CurrencyDto>(currencyResult));
```
Comment (Vlad1slav-Hmeln1tsk1j) [inline CryptoExchange/Microservices/CurrencyManagement/CurrencyManagement.Api/Controllers/CurrenciesController.cs:66]: If something breaks, will everything be Ok?

File: CryptoExchange/Microservices/CurrencyManagement/CurrencyManagement.Api/Controllers/CurrenciesController.cs [hunk 3]
```diff
@@ -58,12 +77,31 @@ public async Task<IActionResult> Update(int id, [FromBody] UpdateCurrencyDto upd
             return Ok(updateCurrencyDto);
```
Comment (Vlad1slav-Hmeln1tsk1j) [inline CryptoExchange/Microservices/CurrencyManagement/CurrencyManagement.Api/Controllers/CurrenciesController.cs:77]: If something goes wrong, everything will still be OK?

File: CryptoExchange/Microservices/CurrencyManagement/CurrencyManagement.Api/Controllers/CurrenciesController.cs [hunk 4]
```diff
@@ -58,12 +77,31 @@ public async Task<IActionResult> Update(int id, [FromBody] UpdateCurrencyDto upd
...
         }
 
+        [HttpPut("IncreasePrice/{id}")]
+        public async Task<IActionResult> IncreasePrice(int id,
+            [FromBody] ChangeCurrencyPriceDto increasePrice)
+        {
+            var currency = await _currencyService.UpdatePriceAsync(id, increasePrice.PriceChange);
+
+            return Ok(_mapper.Map<CurrencyDto>(currency));
```
Comment (Vlad1slav-Hmeln1tsk1j) [inline CryptoExchange/Microservices/CurrencyManagement/CurrencyManagement.Api/Controllers/CurrenciesController.cs:86]: If something goes wrong, everything will still be OK?

File: CryptoExchange/Microservices/CurrencyManagement/CurrencyManagement.Api/Controllers/CurrenciesController.cs [hunk 5]
```diff
@@ -58,12 +77,31 @@ public async Task<IActionResult> Update(int id, [FromBody] UpdateCurrencyDto upd
...
+        }
+
+        [HttpPut("ReducePrice/{id}")]
+        public async Task<IActionResult> ReducePrice(int id,
+            [FromBody] ChangeCurrencyPriceDto reducePrice)
+        {
+            var currency = await _currencyService.UpdatePriceAsync(id, -reducePrice.PriceChange);
+
+            return Ok(_mapper.Map<CurrencyDto>(currency));
```
Comment (Vlad1slav-Hmeln1tsk1j) [inline CryptoExchange/Microservices/CurrencyManagement/CurrencyManagement.Api/Controllers/CurrenciesController.cs:95]: If something goes wrong, everything will still be OK?

File: CryptoExchange/Microservices/UserManagement/UserManagement.Api/UserManagement.Api/Controllers/UsersController.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            _mapper = mapper;
+        }
+
+        [HttpPost("registration")]
+        [AllowAnonymous]
+        public async Task<IActionResult> Registration([FromBody] RegisterUserDto registerUserDto)
+        {
+            var token = await _registrationService.Registration(registerUserDto);
+
+            if (token == null) return BadRequest("User already exists");
```
Comment (Vlad1slav-Hmeln1tsk1j) [inline CryptoExchange/Microservices/UserManagement/UserManagement.Api/UserManagement.Api/Controllers/UsersController.cs]: I don't think it's worth using an if loop without a body

File: CryptoExchange/Microservices/UserManagement/UserManagement.Api/UserManagement.Api/Controllers/UsersController.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+            return Ok(token);
+        }
+
+        [HttpGet]
+        [Authorize(Roles = "Admin")]
+        public async Task<IActionResult> GetAll()
+        {
+            var users = await _userService.GetAllAsync();
+            return Ok(_mapper.Map<IEnumerable<UserDto>>(users));
```
Comment (Vlad1slav-Hmeln1tsk1j) [inline CryptoExchange/Microservices/UserManagement/UserManagement.Api/UserManagement.Api/Controllers/UsersController.cs]: If users will be null that's anyway will be ok?

File: CryptoExchange/Microservices/UserManagement/UserManagement.Api/UserManagement.Api/Controllers/UsersController.cs [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            if (User.IsInRole("User") && user.UserName != User.Identity.Name)
+                return BadRequest();
+
+            user.FirstName = updateUserDto.FirstName;
+            user.LastName = updateUserDto.LastName;
+
+            var updatedUser = await _userService.UpdateAsync(user);
+
+            return Ok(_mapper.Map<UserDto>(updatedUser));
+        }
```
Comment (Vlad1slav-Hmeln1tsk1j) [inline CryptoExchange/Microservices/UserManagement/UserManagement.Api/UserManagement.Api/Controllers/UsersController.cs]: too dirty controller

File: CryptoExchange/Microservices/UserManagement/UserManagement.Api/UserManagement.Api/Program.cs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                Reference = new OpenApiReference
+                {
+                    Type=ReferenceType.SecurityScheme,
+                    Id="Bearer"
+                }
+            },
+            new string[]{}
+        }
+    });
+});
```
Comment (Vlad1slav-Hmeln1tsk1j) [inline CryptoExchange/Microservices/UserManagement/UserManagement.Api/UserManagement.Api/Program.cs]: I think you need to use extension methods

**Review by Vlad1slav-Hmeln1tsk1j — CHANGES_REQUESTED** [review]

**Comment by kvtim** [comment]
Completed exception handling by custom ActionResult

**Comment by kvtim** [comment]
Add validation

**Pull request merged**